Update pipeline

jeremybumsted opened the pull request:
Bring the pipeline configuration up to date by
- moving pipeline into a YAML config from a shell script (pipeline.sh can be removed after verified)
- updating docker-compose to use the correct images
- updating the version to 0.2.2 for linter checks

Added some additional bits (including requested changes)
- Fixed integration tests
- Updated out-of-date docker images used in integration tests so they pass now
- Fixed BATS tests relevant to current functionality, skipping many that will pass in future PRs
- Updated plugin schema

jeremybumsted commented:
This PR will update the pipeline so we have the framework to run the pipeline in BK, but it's expected to fail.

toote commented on .buildkite/pipeline.yml:
```diff
@@ -0,0 +1,40 @@
+  DUMP_ENV: "true"
+  VAULT_DEV_ROOT_TOKEN_ID: "88F4384B-98E9-4AE3-B00C-F55678F89080"
+steps:
+  - label: ":sparkles: Lint"
+    plugins:
+      - docker-compose#v4.5.0:
+          run: lint
```
why not use the linter plugin instead?

toote commented on .buildkite/pipeline.yml:
```diff
@@ -0,0 +1,40 @@
+      - docker-compose#v4.5.0:
+          run: lint
+
+  - label: ":shell: Shellcheck"
+    plugins:
+      - docker-compose#v4.5.0:
+          run: shellcheck
```
why not use the shellcheck plugin instead?

toote commented on plugin.yml:
```diff
@@ -22,7 +22,4 @@ configuration:
   required:
     - server
     - auth
-  additionalProperties: false
-  dependencies:
-    auth: [ method ]
```
why these changes?

jeremybumsted replied:
Fixed this in a more recent commit, we want `auth` to be a required property.

toote commented on tests/auth-tests.bats:
```diff
@@ -2,10 +2,10 @@
-# export SSH_ADD_STUB_DEBUG=/dev/tty
-# export VAULT_STUB_DEBUG=/dev/tty
-# export GIT_STUB_DEBUG=/dev/tty
+export SSH_AGENT_STUB_DEBUG=/dev/tty
+export SSH_ADD_STUB_DEBUG=/dev/tty
+export VAULT_STUB_DEBUG=/dev/tty
+export GIT_STUB_DEBUG=/dev/tty
```
these changes should not be committed to the repository

toote commented on .buildkite/pipeline.yml:
```diff
@@ -0,0 +1,40 @@
+env:
+  BK_QUEUE_NAME: "testqueue" # probably don't need this
+  TESTER_VAULT_VERSION: "1.11.2" # this probably needs to get bumped over time
+  SVC_VAULT_VERSION: "1.11.2" # this probably needs to get bumped over time
+  VAULT_ADDR: "http://vault-svc:8200"
+  BUILDKITE_PIPELINE_SLUG: "my_pipeline"
+  DUMP_ENV: "true"
+  VAULT_DEV_ROOT_TOKEN_ID: "88F4384B-98E9-4AE3-B00C-F55678F89080"
```
if they are only used in a single step, why not add them on that step instead?

jeremybumsted replied:
I tried defining the values a few different ways (most of these values are also defined or have defaults in the docker-compose) with varying degrees of success. Ended up only leaving `VAULT_DEV_ROOT_TOKEN_ID` and `VAULT_ADDR` as globally defined, the rest are defined in the plugin env directly.

toote commented on .buildkite/pipeline.yml:
```diff
@@ -0,0 +1,40 @@
+          run: tests
+          
+  - label: ":vault: :test_tube: Integration Tests"
+    command:  /app/.buildkite/steps/test_envvar.sh
```
it may be a better idea to make this command relative and set the `workdir` in the plugin

toote commented on .buildkite/pipeline.yml:
```diff
@@ -0,0 +1,40 @@
+          - docker-compose-integration.yml
+        run: vault-tester
+        env:
+          - VAULT_ADDR=\${VAULT_ADDR}
+          - BUILDKITE_PLUGIN_VAULT_SECRETS_SERVER=\${VAULT_ADDR}
+          - BUILDKITE_PIPELINE_SLUG=\${BUILDKITE_PIPELINE_SLUG}
+          - BUILDKITE_PLUGIN_VAULT_SECRETS_DUMP_ENV=\${DUMP_ENV}
+          - VAULT_DEV_ROOT_TOKEN_ID=\${VAULT_DEV_ROOT_TOKEN_ID}
+          - BUILDKITE_PLUGIN_VAULT_SECRETS_AUTH_TOKEN=\${VAULT_DEV_ROOT_TOKEN_ID}
+          - SVC_VAULT_VERSION=\${SVC_VAULT_VERSION}
+          - TESTER_VAULT_VERSION=\${TESTER_VAULT_VERSION}
```
Most of these variables are already defined in the docker compose file so I don't even think they are really necessary to specify here

jeremybumsted replied:
Agreed - cleaned this up.

toote reviewed (changes requested):
A few minor suggestions and a big one (with regards to the uncommented debug lines in the tests).

That said, I believe that there is a bigger issue at hand: the testing of the plugin runs a very weird setup with an image we don't know where it comes from. My hunch is that it should be possible to run the test in the same vault image as the only requisite appears to be that vault client is installed... in which case I'm not sure that we need a separate vault service docker container

jeremybumsted commented:
Thanks for the review @toote! I agree with all of your suggestions 😅 I'll get those changes in sometime this week!

> That said, I believe that there is a bigger issue at hand: the testing of the plugin runs a very weird setup with an image we don't know where it comes from. My hunch is that it should be possible to run the test in the same vault image as the only requisite appears to be that vault client is installed... in which case I'm not sure that we need a separate vault service docker container

I believe this image runs a vault server in dev mode, so that the integration tests can be run without the need for an actual vault server. I wanted to address this in a different PR, but it might make more sense to have it dealt with in here too.

toote commented on docker-compose-integration.yml:
```diff
@@ -1,21 +1,22 @@
       - VAULT_TOKEN=${VAULT_DEV_ROOT_TOKEN_ID}
-      - VAULT_ADDR=${VAULT_ADDR}
-      - VPC_PROXY=${VPC_PROXY}
+      - VPC_PROXY=${VPC_PROXY:-}
```
this is the same as:

```suggestion
      - VPC_PROXY
```

toote commented on docker-compose-integration.yml:
```diff
@@ -1,21 +1,22 @@
-       timeout: 10s
-       retries: 5
     environment:
       - VAULT_DEV_ROOT_TOKEN_ID=${VAULT_DEV_ROOT_TOKEN_ID}
```
this is the same as:

```suggestion
      - VAULT_DEV_ROOT_TOKEN_ID
```

toote commented on docker-compose-integration.yml:
```diff
@@ -26,13 +27,13 @@ services:
-      - VAULT_ADDR=${VAULT_ADDR}
-      - VPC_PROXY=${VPC_PROXY}
+      - VAULT_DEV_ROOT_TOKEN_ID
+      - VAULT_ADDR=${VAULT_ADDR:-}
+      - VPC_PROXY=${VPC_PROXY:-}
```
this is the same as:
```suggestion
      - VAULT_ADDR
      - VPC_PROXY
```

toote commented on lib/shared.bash:
```diff
@@ -1,11 +1,11 @@
 set -ueo pipefail
 
-BASE64_DECODE_ARGS="-d"
+# BASE64_DECODE_ARGS="-d"
```
if the code is no longer used, remove it

toote commented on lib/shared.bash:
```diff
@@ -1,11 +1,11 @@
-case "$(uname -s)" in
-  Darwin) BASE64_DECODE_ARGS="--decode" ;;
-esac
+# case "$(uname -s)" in
+#   Darwin) BASE64_DECODE_ARGS="--decode" ;;
+# esac
```
same here, if it is no longer used, remove it

toote commented on tests/auth-tests.bats:
```diff
@@ -7,95 +7,46 @@ load '/usr/local/lib/bats/load.bash'
-  assert_output --partial 'Successfully authenticated. You are now logged in'
-
-  unstub vault
+  # assert_output --partial 'Successfully authenticated with RoleID'
```
why did you make these changes?

* All stubbed commands should be unstubbed to ensure they were called as expected
* you are only testing that the hooks did not fail/explode, but no actual functionality otherwise

jeremybumsted replied:
These and the other tests are fixed

toote commented on tests/auth-tests.bats:
```diff
@@ -7,95 +7,46 @@ load '/usr/local/lib/bats/load.bash'
+  # assert_output --partial 'Successfully authenticated with RoleID'
 
   unset BUILDKITE_PLUGIN_VAULT_SECRETS_PATH
   unset BUILDKITE_PLUGIN_VAULT_SECRETS_SERVER
   unset BUILDKITE_PLUGIN_VAULT_SECRETS_DUMP_ENV
-  unset BUILDKITE_PLUGIN_VAULT_SECRETS_AUTH_HEADER
   unset BUILDKITE_PIPELINE_SLUG
+  unset BUILDKITE_PLUGIN_VAULT_SECRETS_AUTH_SECRET_ENV
+  unset BUILDKITE_PLUGIN_VAULT_SECRETS_AUTH_ROLE_ID
+  unset BUILDKITE_PLUGIN _VAULT_SECRETS_AUTH_SECRET_ID
+  unset BUILKDITE_PLGUIN_VAULT_SECRETS_AUTH_METHOD
```
note that it has long been unnecessary to unset all variables exported inside a test, they are all isolated from each other

jeremybumsted replied:
TIL! Updated to remove the unsetting in all the tests 🙌

toote commented on tests/environment-hook.bats:
```diff
@@ -13,14 +13,17 @@ load '/usr/local/lib/bats/load.bash'
-    "list -address=https://vault_svr_url -format=yaml data/buildkite : echo -e '- env'" \
-    "list -address=https://vault_svr_url -format=yaml data/buildkite/env : echo -e '- MY_SECRET'" \
-    "read -address=https://vault_svr_url -field=value data/buildkite/env/MY_SECRET : echo ${TESTDATA}"
+    "write -field=token -address=https://vault_svr_url auth/approle/login role_id=buildkite secret_id=\"${BUILDKITE_PLUGIN_VAULT_SECRETS_AUTH_SECRET_ENV}\" : echo 'Successfully authenticated with Role ID ${BUILDKITE_PLUGIN_VAULT_SECRETS_AUTH_ROLE_ID}'" \
```
You can make this stub more generic (and shorter) with the following:

```suggestion
    "write -field=token \* auth/approle/login \* \* : echo 'Successfully authenticated with \${5} \${6}" \
```

Which is basically accepting any parameters in the 3rd, 5th and 6th positions and using the last two in the output so you can later make assertions on the output

jeremybumsted replied:
I've actually completely removed the auth section of these tests because it's not needed to test the functionality of evaluating `env` received from Vault. This is good to know though!

toote commented on tests/environment-hook.bats:
```diff
@@ -13,14 +13,17 @@ load '/usr/local/lib/bats/load.bash'
+    "write -field=token -address=https://vault_svr_url auth/approle/login role_id=buildkite secret_id=\"${BUILDKITE_PLUGIN_VAULT_SECRETS_AUTH_SECRET_ENV}\" : echo 'Successfully authenticated with Role ID ${BUILDKITE_PLUGIN_VAULT_SECRETS_AUTH_ROLE_ID}'" \
+    "kv list -address=https://vault_svr_url -format=yaml data/buildkite/testpipe : echo 'env'" \
+    "kv list -address=https://vault_svr_url -format=yaml data/buildkite : echo 'env'" \
+    "kv get -address=https://vault_svr_url -field=data -format=yaml data/buildkite/testpipe/env : echo ${TESTDATA}" \
+    "kv get -address=https://vault_svr_url -field=data -format=yaml data/buildkite/env : echo ${TESTDATA}"
```
You are outputting the same string twice, how can you distinguish them in the output? (i.e.: if one of these is not called, the `assert_output` would not fail). I understand that other assertions or the unstub most likely would, but if you are testing some functionality, it may be a good idea to test it all.

toote commented on tests/environment-hook.bats:
```diff
@@ -87,26 +104,30 @@ load '/usr/local/lib/bats/load.bash'
 #-------
 # Project scope
 @test "Load project env file from vault server" {
+  skip "this test won't pass completely, for now"
```
can you add more information as to why?

toote commented on tests/environment-hook.bats:
```diff
@@ -123,17 +144,18 @@ load '/usr/local/lib/bats/load.bash'
 }
 
 @test "Load project environment file from vault server" {
+  skip "this test won't pass completely, for now"
```
can you add more information as to why?

toote commented on tests/environment-hook.bats:
```diff
@@ -150,20 +172,21 @@ load '/usr/local/lib/bats/load.bash'
 }
 
 @test "Load project env and environments files from vault server" {
+  skip "this test won't pass completely, for now"
```
can you add more information as to why?

toote reviewed (changes requested):
Great work! A lot of comments here and there, though.

The most concerning to me are:
* plugins in a single step
* skipping of tests (if you are skipping a test indicate exactly why so that anyone can turn them on afterwards)

jeremybumsted commented on .buildkite/steps/vault-init-tester.sh:
```diff
@@ -1,16 +1,19 @@
+TESTDATA_1="foobar1"
+TESTDATA_2="foobar2"
 
 [ $? -ne 1 ] && {
```
@toote I think this is valid since it's not a variable assignment but rather an operation on the vault server. The last command run is `vault login token="$AUTH_TOKEN"` which will authenticate to the vault service that is spun up as part of the integration tests using the `AUTH_TOKEN`, it tests that the script is able to login to `vault` with the supplied token and enable the `data/buildkite` kv2 secrets engine. There might be a more efficient way of doing this though.

toote replied:
What I'm saying is that the value of `$?` (the exit code of the command in line 9) gets overwritten on every single command run. In particular this is the value of the variable for each line:

* line 11: if `vault login` failed or not
* line 13: 
  - if it was `0` in line 11, it is now the result of the `vault secrets enable`
  - otherwise, it will remain unchanged
* line 14: `0` (because the variable assignment of line 13 can not fail)
* live 16: `0` (because the variable assignment of line 14 can not fail)

I double-checked this:
```
$ bash -c 'exit 1'
$ echo $?
1
$ bash -c 'exit 1'
$ [ $? -eq 1 ] && echo true
true
$ echo $?
0
$ bash -c 'exit 1'
$ [ $? -eq 0 ] && echo true
$ echo $?
1
$ bash -c 'exit 1'
$ FOO=10
$ echo $?
0
```

Based on what I believe it is attempted here, and avoiding the use and abuse of `$?` and `test` in the form of `[]`, the script should be something like:

```
if ! vault login ...; then
  echo 'was not able to log in'
  exit 1
fi

if ! vault mount -path data/buildkite kv; then
  echo 'was not able to mount'
  exit 2 # just to be different from the login error
fi

vault kv put data..
```

jeremybumsted replied:
Ahhh, this is where my knowledge of bash is pretty weak so I've learned some new things here. I've tested your suggestions and things are working as expected, thanks for that explanation!

jeremybumsted commented on Dockerfile:
```diff
@@ -0,0 +1,3 @@
+FROM bash:alpine3.16
+
+RUN apk add --no-cache vault
```
Flipped over to using the `vault` image in the Dockerfile and adding bash to the container so we can take advantage of running the tests inside of bash instead of Alpine's default shell

jeremybumsted commented on .buildkite/steps/test_envvar.sh:
```diff
@@ -1,42 +1,44 @@
 setupTestData() {
   export VAULT_TOKEN="${VAULT_DEV_ROOT_TOKEN_ID:-}"
-  ${basedir}/.buildkite/steps/vault-init-tester.sh
+  sh "${basedir}/.buildkite/steps/vault-init-tester.sh"
```
Fixed this by switching over to the official `vault` image and adding `bash`

jeremybumsted replied:
Ah, this was a mistake 😓 I switched this up trying to get the script to run in the integration-test container, I've reverted back to running in the same shell now.

jeremybumsted commented:
@toote 👋 Ended up going in and just fixing all the tests so everything is working now and passing 👍

toote commented on .buildkite/steps/test_envvar.sh:
```diff
@@ -1,42 +1,44 @@
 configEnv() {
-  _source="${BASH_SOURCE[0]}"
+  _source="$0"
   [ -z "${_source:-}" ] && _source="${0}"
```
still not sure why line 7 is checking if a variable that you just defined is still defined and if not, defines it the exact same way as you did before

jeremybumsted replied:
Yeah, doesn't make much sense... I've updated it as such:

```shell
configEnv() {
  basedir="$( cd "$( dirname "$0" )" && cd ../.. && pwd )"
}
```

toote commented on .buildkite/steps/test_envvar.sh:
```diff
@@ -1,42 +1,44 @@
 }
 
 setupTestData() {
   export VAULT_TOKEN="${VAULT_DEV_ROOT_TOKEN_ID:-}"
```
still not sure what this export is for

jeremybumsted replied:
Yeah, not needed - removed.

toote commented on .buildkite/steps/test_envvar.sh:
```diff
@@ -1,42 +1,44 @@
+
 
-  if [[ ! -z "${TESTDATA_1:-}" ]] && [[ "${TESTDATA_1}" == "foo bar 1" ]] ; then
+  if [[ -n "${TESTDATA_1:-}" ]] && [[ "${TESTDATA_1}" == "foobar1" ]] ; then
```
my previous comment still applies

toote commented on .buildkite/steps/test_envvar.sh:
```diff
@@ -1,42 +1,44 @@
   fi
 
-  if [[ ! -z "${TESTDATA_2:-}" ]] && [[ "${TESTDATA_2}" == "foo bar 2" ]] ; then
+  if [[ ! -z "${TESTDATA_2:-}" ]] && [[ "${TESTDATA_2}" == "foobar2" ]] ; then
```
my previous comment still applies here

toote commented on .buildkite/steps/vault-init-tester.sh:
```diff
@@ -1,16 +1,19 @@
 AUTH_TOKEN="${BUILDKITE_PLUGIN_VAULT_SECRETS_AUTH_TOKEN:-}"
 PROJECT=${BUILDKITE_PIPELINE_SLUG:-foobar_project}
-vault auth - <<< "$AUTH_TOKEN"
+export VAULT_ADDR="http://vault-svc:8200"
```
my previous comment still applies here

jeremybumsted replied:
This one is a bit weird... I've tried a couple of different methods to make sure things work appropriately, but basically if no value is set, `VAULT_ADDR` will default to `http://localhost:8200` which will cause the integration tests to fail. I've managed to work around this though, so this is fixed.

toote commented on tests/environment-hook.bats:
```diff
@@ -29,25 +31,22 @@ load '/usr/local/lib/bats/load.bash'
-    "list -address=https://vault_svr_url -format=yaml data/buildkite : echo -e '- environment'" \
-    "list -address=https://vault_svr_url -format=yaml data/buildkite/environment : echo -e '- MY_SECRET'" \
-    "read -address=https://vault_svr_url -field=value data/buildkite/environment/MY_SECRET : echo ${TESTDATA}"
+    "write -field=token -address=https://vault_svr_url auth/approle/login role_id=buildkite secret_id=\"${BUILDKITE_PLUGIN_VAULT_SECRETS_AUTH_SECRET_ENV}\" : echo 'Successfully authenticated with Role ID ${BUILDKITE_PLUGIN_VAULT_SECRETS_AUTH_ROLE_ID}'" \
```
you can reuse the same generic stub as before to make it more general and less verbose:

```suggestion
    "write -field=token \* auth/approle/login \* \* : echo 'Successfully authenticated with \${5} \${6}" \
```

jeremybumsted replied:
Both this and the other tests which had an auth method [configured ](https://github.com/buildkite-plugins/vault-secrets-buildkite-plugin/pull/13#discussion_r1111105758) didn't actually need them to be defined to test the functionality, so I've removed this stub

toote commented on tests/environment-hook.bats:
```diff
@@ -56,28 +55,26 @@ load '/usr/local/lib/bats/load.bash'
-    "read -address=https://vault_svr_url -field=value data/buildkite/env/MY_SECRET : echo ${TESTDATA_ENV1}" \
-    "list -address=https://vault_svr_url -format=yaml data/buildkite/environment : echo -e '- ANOTHER_SECRET'" \
-    "read -address=https://vault_svr_url -field=value data/buildkite/environment/ANOTHER_SECRET : echo ${TESTDATA_ENV2}"
+    "write -field=token -address=https://vault_svr_url auth/approle/login role_id=buildkite secret_id=\"${BUILDKITE_PLUGIN_VAULT_SECRETS_AUTH_SECRET_ENV}\" : echo 'Successfully authenticated with Role ID ${BUILDKITE_PLUGIN_VAULT_SECRETS_AUTH_ROLE_ID}'" \
```
same here:
```suggestion
    "write -field=token \* auth/approle/login \* \* : echo 'Successfully authenticated with \${5} \${6}" \
```

toote reviewed (changes requested):
Still several comments of my previous review remain unaddressed and I found a few more things to check. Finally, the tests are still not passing on the latest plugin tester version (v4.0.0), luckily it is a simple fix in the ssh tests (missing a `*` in the shh-keygen stub of the first test)

jeremybumsted commented on .buildkite/pipeline.yml:
```diff
@@ -0,0 +1,34 @@
+          run: tests
+          
+  - label: ":vault: :test_tube: Integration Tests"
+    command: bash .buildkite/steps/test_envvar.sh
```
This was an artifact of the old images I was using, no longer needed in the current configuration, so removed

jeremybumsted commented:
@toote Thanks again for the review! I've addressed your comments, as well as fixed some unintended behaviour in the auth function that wasn't using the secret ID properly (and causing some tests to fail inadvertently).

toote reviewed (approved):
:clap: awesome work!
